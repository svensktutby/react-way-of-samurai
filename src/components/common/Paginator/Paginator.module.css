.btn {
  display: inline-block;
  padding: 4px 9px;

  vertical-align: middle;
  color: #7d97a5;

  background-color: #fff;
  border: 1px solid #7d97a5;
  cursor: pointer;

  transition-duration: 0.5s;
  transition-property: color, background-color, opacity, box-shadow;

  user-select: none;
}

.btn:not(:last-child) {
  border-right: 0;
}

.btn:first-child {
  border-top-left-radius: 3px;
  border-bottom-left-radius: 3px;
}

.btn:last-child {
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
}

.btn:not(.selected):disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

@media (hover: hover) {
  .btn:not(:disabled):hover,
  .btn:focus {
    background-color: rgba(125, 151, 165, 0.4);
    outline: 0;
  }
}

.selected,
.selected:hover {
  font-weight: bold;
  color: #fff;

  background-color: #7d97a5;
  cursor: default;
}
