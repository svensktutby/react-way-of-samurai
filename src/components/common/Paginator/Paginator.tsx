import React, { FC } from 'react';

import s from './Paginator.module.css';

type PaginatorPropsType = {
  pageSize: number;
  totalItemsCount: number;
  currentPage?: number;
  changePageHandler: (page: number) => void;
  portionSize?: number;
};

export const Paginator: FC<PaginatorPropsType> = ({
  pageSize,
  totalItemsCount,
  currentPage = 1,
  changePageHandler,
  portionSize = 10,
}) => {
  const portionCount = Math.ceil(totalItemsCount / pageSize);

  const startIndex = Math.floor((currentPage - 1) / portionSize) * portionSize;

  const pages = Array(portionSize)
    .fill(null)
    .reduce((acc, item, idx) => {
      const current = idx + 1 + startIndex;

      if (current <= portionCount) {
        return [...acc, current];
      }

      return acc;
    }, []);

  return (
    <div className={s.paginator}>
      <button
        type="button"
        className={`${s.btn}`}
        disabled={currentPage === 1}
        onClick={() => changePageHandler(1)}
      >
        First
      </button>
      <button
        type="button"
        className={`${s.btn}`}
        disabled={currentPage === 1}
        onClick={() => changePageHandler(currentPage - 1)}
      >
        Previous
      </button>
      {pages.map((p: number) => {
        const isCurrent = p === currentPage;

        return (
          <button
            type="button"
            key={p}
            className={`${s.btn} ${isCurrent && s.selected}`}
            disabled={isCurrent}
            onClick={() => changePageHandler(p)}
          >
            {p}
          </button>
        );
      })}
      <button
        type="button"
        className={`${s.btn}`}
        disabled={currentPage === portionCount}
        onClick={() => changePageHandler(currentPage + 1)}
      >
        Next
      </button>
      <button
        type="button"
        className={`${s.btn}`}
        disabled={currentPage === portionCount}
        onClick={() => changePageHandler(portionCount)}
      >
        Last
      </button>
    </div>
  );
};
