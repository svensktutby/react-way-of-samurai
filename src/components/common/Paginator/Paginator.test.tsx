import React from 'react';
import TestRenderer, { ReactTestInstance } from 'react-test-renderer';

import { Paginator } from './Paginator';

let root: ReactTestInstance;
const mockCallback = jest.fn();
const props = {
  pageSize: 5,
  totalItemsCount: 100,
  currentPage: 1,
  portionSize: 10,
  changePageHandler: mockCallback,
};

describe('Paginator component', () => {
  beforeEach(() => {
    root = TestRenderer.create(<Paginator {...props} />).root;
  });

  it('should matches the snapshot', () => {
    const component = TestRenderer.create(<Paginator {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render all buttons', () => {
    const buttons = root.findAllByType('button');
    expect(buttons).toHaveLength(14);
  });

  it('callback should be called', () => {
    root.props.changePageHandler(2);
    expect(mockCallback.mock.calls).toHaveLength(1);
  });
});
