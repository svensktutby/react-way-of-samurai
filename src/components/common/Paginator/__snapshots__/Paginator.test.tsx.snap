// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator component should matches the snapshot 1`] = `
<div
  className="paginator"
>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    First
  </button>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Previous
  </button>
  <button
    className="btn selected"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className="btn false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className="btn false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="btn false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    className="btn false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    className="btn false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    6
  </button>
  <button
    className="btn false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    7
  </button>
  <button
    className="btn false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    8
  </button>
  <button
    className="btn false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    9
  </button>
  <button
    className="btn false"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    10
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Last
  </button>
</div>
`;
